openid_connect.js

```diff
@@ -7,7 +7,38 @@ var newSession = false; // Used by oidcAuth() and validateIdToken()
 
 export default {auth, codeExchange, validateIdToken, logout};
 
+function sleep(ms) {
+    return new Promise(resolve => setTimeout(resolve, ms));
+}
+
+async function isVarDefined(variable) {
+    const v = await sleep(10);
+    if (variable != null) return true;
+    else return false;
+}
+
+// If the ID token has not been synced yet, poll the variable every 10ms until
+// get a value or after a timeout.
+async function waitUntilVarIsDefined(r, timeout) {
+    let timeWas = new Date();
+    let result = false;
+    do {
+        result = await isVarDefined(r.variables.session_jwt);
+    } while ((! result) && (new Date() - timeWas < timeout));
+    return result;
+}
+
 function auth(r) {
+    // If a cookie was sent but the ID token is not in the key-value database, wait for the token to be in sync.
+    if (r.variables.cookie_auth_token != null && r.variables.session_jwt == null && r.variables.zone_sync_timeout > 0) {
```

Comment:
Shall we also check the presence of `refresh_token` or perhaps move all this logic under related `if` below?

Reply:
I don't think it makes much sense. I can't imagine a situation where we have refresh token in kvs, but we don't have an ID token. But even if such a situation takes place, it doesn't break anything, after the "zone_sync_leeway" time we'll get a new set of tokens and make an internal redirect to the same location, so everything will work from the end user's point of view.

Reply:
The situation is ID token expiry in keyval (and by default it's 1h). I don't see why user has to wait for zone_sync_leeway in that case.